aesd-char-driver/main.c

```diff
@@ -48,23 +55,99 @@ ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
                 loff_t *f_pos)
 {
 	ssize_t retval = 0;
+	struct aesd_dev *chardev = filp->private_data;
+        struct aesd_buffer_entry *entry = NULL;
+        size_t entry_offset = 0;
+        size_t num_bytes_to_read = 0;
 	PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
 	/**
 	 * TODO: handle read
 	 */
+
+	if (mutex_lock_interruptible(&chardev->lock))
+		return -ERESTARTSYS;
+	
+	entry = aesd_circular_buffer_find_entry_offset_for_fpos(&chardev->cb, *f_pos, &entry_offset);
+
+	if(entry == NULL)
+	{
+		retval = 0;
+		goto out;
+	}
+
+	num_bytes_to_read = entry->size - entry_offset;
+	if(num_bytes_to_read > count)
+	{
+		num_bytes_to_read = count;
+	}
+
+	if((copy_to_user(buf, entry->buffptr + entry_offset, num_bytes_to_read))) 
+	{
+		retval = -EFAULT;
+		goto out;
+	}
+
+	*f_pos += num_bytes_to_read;
+	
+	mutex_unlock(&chardev->lock);
+	return num_bytes_to_read;
+out:
+	mutex_unlock(&chardev->lock);
```

Comment:
@midhun9 you should be able to remove the duplicate return here if you just set retval with `num_bytes_to_read`